async-server.c

```diff
@@ -0,0 +1,727 @@
+
+    for (int i = 0; i < size; ++i)
+    {
+        set->data[i] = i;
+    }
+    set->capacity = size;
+    set->size = size;
+}
+
+static long
+borrow_id(struct id_set_t *set)
+{
+    sem_wait(&set->available);
+    long id = set->data[0];
+    --set->size;
+    set->data[0] = set->data[set->size];
+    return id;
+}
+
+static void
+return_id(struct id_set_t *set, long id)
+{
+    set->data[set->size] = id;
+    ++set->size;
+    sem_post(&set->available);
+}
+
+static void
+id_set_destroy(struct id_set_t *set)
+{
+    free(set->data);
+}
+
+struct srv_context_t
+{
+    volatile int tasks_running;
+    pthread_mutex_t tasks_mutex;
+    pthread_cond_t tasks_cond;
+
+    struct set_t set;
+    pthread_mutex_t set_mutex;
+    struct queue_t queue;
+    password_t password;
+    char *hash;
+    volatile bool found;
+
+    pthread_mutex_t thread_started;
+
+    struct config_t *config;
+};
+
+struct handler_context_t
+{
+    int max_tasks;
+    int current_tasks;
+    struct task_t *tasks;
+    struct id_set_t id_set;
+    pthread_mutex_t id_set_mutex;
+
+    int client_sfd;
+    struct srv_context_t *srv_context;
+};
+
+static void
+handler_context_destroy(struct handler_context_t *context)
+{
+    free(context->tasks);
+    id_set_destroy(&context->id_set);
+    pthread_mutex_destroy(&context->id_set_mutex);
+}
+
+struct params_t
+{
+    struct srv_context_t *context;
+    int socket_fd;
+};
+
+static int
+send_tag(int socket_fd, enum command_t tag)
+{
+    struct iovec vec[1];
+    vec[0].iov_base = &tag;
+    vec[0].iov_len = sizeof(tag);
+    return sendall_vec(socket_fd, vec, 1);
+}
+
+static int
+send_message(int socket_fd, enum command_t tag, void *data, int length)
+{
+    struct iovec vec[3];
+    vec[0].iov_base = &tag;
+    vec[0].iov_len = sizeof(tag);
+
+    vec[1].iov_base = &length;
+    vec[1].iov_len = sizeof(length);
+
+    vec[2].iov_base = data;
+    vec[2].iov_len = length;
+
+    int status = sendall_vec(socket_fd, vec, sizeof(vec) / sizeof(vec[0]));
+    if (status == -1) return -1;
+
+    return 0;
+}
+
+static int
+close_client(int client_sfd)
+{
+    int status;
+
+    fprintf(stderr, "Sending EXIT\n");
+    status = send_tag(client_sfd, CMD_EXIT);
+    if (status == -1) return -1;
+
+    // Client should close socket on its side and send EOF
+    char res;
+    status = recv(client_sfd, &res, sizeof(res), 0);
+
+    shutdown(client_sfd, SHUT_RDWR);
+    close(client_sfd);
+    return 0;
+}
+
+static void *
+task_sender(void *arg)
+{
+    struct handler_context_t *context = (struct handler_context_t *) arg;
+    struct srv_context_t *srv_context = context->srv_context;
+    int client_sfd = context->client_sfd;
+
+    int status;
+
+    while (true)
+    {
+        if (srv_context->tasks_running == 0)
+        {
+            fprintf(stderr, "No tasks left, exiting sender\n");
+            break;
+        }
+        else
+        {
+            long id = borrow_id(&context->id_set);
```

Comment:
Here is a race condition with return_id in task_receiver. You need a lock for access to this set